xbmc/Application.cpp

```diff
@@ -3232,11 +3225,27 @@ PlayBackRet CApplication::PlayFile(CFileItem item, const std::string& player, bo
 
   // We should restart the player, unless the previous and next tracks are using
   // one of the players that allows gapless playback (paplayer, VideoPlayer)
-  // DVD playback does not support gapless
-  if (item.IsDiscImage() || item.IsDVDFile())
-    m_appPlayer.ClosePlayer();
-  else
-    m_appPlayer.ClosePlayerGapless(newPlayer);
+  if (m_appPlayer.HasPlayer())
+  {
+    if (m_appPlayer.CanClosePlayerGapless(newPlayer, item))
+    {
+      m_appPlayer.ClosePlayerGapless(newPlayer);
+    }
+    else
+    {
+      m_appPlayer.ClosePlayer();
+      if (m_itemCurrentFile != nullptr)
+      {
+        // wait for m_itemCurrentFile being reset to nullptr (asynch job)
```

Comment:
In general PlayFile is run on the main thread and this must not be blocked. I whole idea od asynchronous processing is not blocking threads. Instead of waiting here, you should hook into OnPlayerCloseFile

Reply:
If you read the description, you'll notice that this is to deal with the situation of two players (non-gapless). The new player starts before the old one can stop. Both are calling OnXY methods, and that's a mess you need to synchronize!

Reply:
blocking mainthread is no-go. 
why not starting the new player when the old one finishes?

Reply:
So we are in PlayFile here, waiting to start the (new) file. Before we can do this, we must make sure the old player finishes ALL work (including the asynch stuff in OnPlaybackStopped/OnPlaybackEnded). So you need to wait here, otherwise CreatePlayer() spawns a new player thread right after this. I'm open to suggestions on how to wait. Note that you can't count on it being the same player class. E.g. you can jump between VP and PAP...

Reply:
Instead of waiting here CApp can switch to a waiting state and exit this method. When OnPlybackEnded/Stopped is called and the app is in this waiting state, it can eventually start new player. We may need to cut PlayFile into smaller chunks and execute those chunks on events.

Reply:
Seems very complicated given current situation. Here's another idea: how about adding the player thread ID as a parameter to OnPlaybackEnded/Stopped ? Then in OnPlaybackStarted, I would keep track of m_currentFileItem associated to player thread. Could easily hold them in a map < thread id, current item > . Minimal change...

Reply:
No hacks, lets do this right. I don't think my suggestion is complicated. If you want me to do this part, let me know.

Reply:
okay, that's greatly appreciated. I'll help where I can (e.g. testing)

Reply:
ok, but needs to wait until tomorrow. this evening is Jedi time :)